### PR TITLE
[jaeger-v2] Streamline storage initialization

### PR DESCRIPTION
**Pull request by yurishkuro**
## Which problem is this PR solving?
- There is duplicated code in the extension Start, which is being copy/pasted in other PRs

## Description of the changes
- Refactor Start to avoid duplication and to make testing simpler (no need to test all edge cases for every storage type)
- Clean-up tests

## How was this change tested?
- 

## Checklist
- [ ] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [ ] I have signed all commits
- [ ] I have added unit tests for the new functionality
- [ ] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

### CODE REVIEW

File: cmd/jaeger/internal/extension/jaegerstorage/extension.go [hunk 1]
```diff
@@ -60,35 +62,62 @@ func newStorageExt(config *Config, otel component.TelemetrySettings) *storageExt
...
+	return nil
+}
 
-	// TODO add support for other backends
+func (s *storageExt) Start(ctx context.Context, host component.Host) error {
+	memStarter := &starter[memoryCfg.Configuration, *memory.Factory]{
+		ext:         s,
+		storageKind: "memory",
+		cfg:         s.config.Memory,
+		// memory factory does not return an error, so need to wrap it
```
Comment (albertteoh) [inline cmd/jaeger/internal/extension/jaegerstorage/extension.go:95]: Why not change `memory.NewFactoryWithConfig` to return the error? It doesn't look like it's used anywhere else, so it shouldn't be a large breaking change.

File: cmd/jaeger/internal/extension/jaegerstorage/extension.go [hunk 2]
```diff
@@ -60,35 +62,62 @@ func newStorageExt(config *Config, otel component.TelemetrySettings) *storageExt
...
+	ext         *storageExt
+	storageKind string
+	cfg         map[string]Config
+	builder     func(Config, metrics.Factory, *zap.Logger) (Factory, error)
+}
 
-	for name, b := range s.config.Badger {
-		if _, ok := s.factories[name]; ok {
-			return fmt.Errorf("duplicate badger storage name %s", name)
+func (s *starter[Config, Factory]) build(ctx context.Context, host component.Host) error {
```
Comment (albertteoh) [inline cmd/jaeger/internal/extension/jaegerstorage/extension.go:72]: Internal state isn't persisted between method calls so all methods are effectively read-only; we could assert this by using value receivers instead of pointer receivers.

**Review by albertteoh — APPROVED** [review]

File: cmd/jaeger/internal/extension/jaegerstorage/extension.go (hunk 1 above)
Comment (yurishkuro) [reply cmd/jaeger/internal/extension/jaegerstorage/extension.go:95]: Because I hate functions that are declared with errors but always return nil. The caller either has to ignore the error, or it has no easy way of unit testing the error handing path (which will never be used). Lose-lose. I think if the function can never return an error it should never declare it as a possibility (unless you're dealing with an interface where there's no choice).

File: cmd/jaeger/internal/extension/jaegerstorage/extension.go (hunk 2 above)
Comment (yurishkuro) [reply cmd/jaeger/internal/extension/jaegerstorage/extension.go:72]: true, but makes no difference in this context since the code runs exactly once. Don't want to reset the approval just for this fix :-)
